File: Unique_Paths_II.java
/**
 * Problem Link: https://leetcode.com/problems/unique-paths-ii/
 */
public class Solution {
    public int uniquePathsWithObstacles(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int[][] cost = new int[n][m];
        cost[0][0] = grid[0][0] == 0 ? 1 : 0;
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(grid[i][j] == 0) {
                    if(i > 0) cost[i][j] += cost[i-1][j];
                    if(j > 0) cost[i][j] += cost[i][j-1];
                }
                else {
                    cost[i][j] = 0;
                }
            }
        }
        return cost[n-1][m-1];
    }
}

File: Swift/unique-paths-ii.swift
/**
 * https://leetcode.com/problems/unique-paths-ii/
 * 
 * 
 */ 
class Solution {
    func uniquePathsWithObstacles(_ obstacleGrid: [[Int]]) -> Int {
        let n = obstacleGrid.count
        guard let m = obstacleGrid.first?.count else { return 0 }
        var mark = Array(repeating: Array(repeating: 0, count: m), count: n)
        mark[0][0] = 1 - obstacleGrid[0][0]
        for x in 0 ..< n {
            for y in 0 ..< m {
                if obstacleGrid[x][y] == 1 { continue }
                if x > 0 {
                    mark[x][y] += mark[x - 1][y] 
                }
                if y > 0 {
                    mark[x][y] += mark[x][y - 1]
                }
            }
        }
        return mark[n - 1][m - 1]
    }
}